### PR TITLE
[BEAM-121] Add DisplayData for IO transforms

### PR DESCRIPTION
**Pull request by swegner**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace "<Jira issue #>" in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by swegner** [comment]
R=@dhalperi @bjchambers

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/AvroIO.java [hunk 1]
```diff
@@ -326,6 +327,14 @@
       }
 
       @Override
+      public void populateDisplayData(DisplayData.Builder builder) {
+        builder.add("schema", type);
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/AvroIO.java]: What happens here? `type` is a `Class`. Will it just get automatically converted to a `String`?

I would probably just drop the type. We aren't planning to specify the `Coder` for each transform as static metadata, are we?

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/AvroIO.java [hunk 2]
```diff
@@ -326,6 +327,14 @@
...
+        if (filepattern != null) {
+          builder.add("filePattern", filepattern);
+        }
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/AvroIO.java]: how do you want to handle the properties for nested transforms?

Specifically, filepattern is really just passed directly through to `FileBasedSource` -- should it be there instead?

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/AvroIO.java [hunk 3]
```diff
@@ -678,6 +687,22 @@ public PDone apply(PCollection<T> input) {
...
 
+      @Override
+      public void populateDisplayData(DisplayData.Builder builder) {
+        builder.add("schema", type);
+        if (filenamePrefix != null) {
+          builder.add("fileNamePrefix", filenamePrefix);
+        }
+
+        if (filenameSuffix != null && !filenameSuffix.isEmpty()) {
+          builder.add("fileNameSuffix", filenameSuffix);
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/AvroIO.java]: Same question re `FileBasedSink`

Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/AvroIO.java]: This is incomplete -- look at every function that returns a `new Bound`, Missing shard name template, etc.

I wonder if, in `FileBasedSink`, we could just assemble the entire filename (template) and use that.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/AvroSource.java [hunk 4]
```diff
@@ -293,6 +294,20 @@ public boolean producesSortedKeys(PipelineOptions options) throws Exception {
...
+
+    if (readSchemaString != null) {
+      builder.add("schema", readSchemaString);
+    }
+
+    long minBundleSize = getMinBundleSize();
+    if (minBundleSize != DEFAULT_MIN_BUNDLE_SIZE) {
+      builder.add("minBundleSize", minBundleSize);
+    }
+  }
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/AvroSource.java]: see comments on AvroIO.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java [hunk 5]
```diff
@@ -501,6 +502,21 @@ private static void dryRunQuery(BigQueryOptions options, String query) {
         return TableRowJsonCoder.of();
       }
 
+      @Override
+      public void populateDisplayData(DisplayData.Builder builder) {
+        if (table != null) {
+          builder.add("tableSpec", toTableSpec(table));
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java]: `table`

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java [hunk 6]
```diff
@@ -501,6 +502,21 @@ private static void dryRunQuery(BigQueryOptions options, String query) {
...
+        }
+
+        if (query != null) {
+          builder.add("query", query);
+        }
+
+        if (flattenResults != null) {
+          builder.add("flattenResults", flattenResults);
+        }
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java]: How do you want to handle options with default values? I would probably have done:

```java
if (!validation) {
   builder.add("validation", validation);
}
```
to only show this if validation is disabled.

(Or if there is a null-valued property where simply presence/absence of key is meaningful, just "validation disabled").

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java [hunk 7]
```diff
@@ -970,6 +986,21 @@ public PDone apply(PCollection<TableRow> input) {
         return VoidCoder.of();
       }
 
+      @Override
+      public void populateDisplayData(DisplayData.Builder builder) {
+        if (table != null) {
+          builder.add("tableSpec", toTableSpec(table));
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java]: table

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java [hunk 8]
```diff
@@ -970,6 +986,21 @@ public PDone apply(PCollection<TableRow> input) {
...
+        }
+
+        if (schema != null) {
+          builder.add("schema", schema.toString());
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java]: this will be a JSON blob. Should it be prettified? Should we have monitoring support for prettification? any size limits?

(probably good as is)

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java [hunk 9]
```diff
@@ -970,6 +986,21 @@ public PDone apply(PCollection<TableRow> input) {
...
+        }
+
+        builder
+            .add("createDisposition", createDisposition.toString())
+            .add("writeDisposition", writeDisposition.toString());
+      }
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java]: same possible concern about validation.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java [hunk 10]
```diff
@@ -970,6 +986,21 @@ public PDone apply(PCollection<TableRow> input) {
         return VoidCoder.of();
       }
 
+      @Override
+      public void populateDisplayData(DisplayData.Builder builder) {
+        if (table != null) {
+          builder.add("tableSpec", toTableSpec(table));
+        }
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java]: if `tableSpecFn` is present, probably worth adding `table function: tableSpecFn.getClass().getSimpleName()`.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/BoundedReadFromUnboundedSource.java [hunk 11]
```diff
@@ -107,6 +108,21 @@ public String getKindString() {
...
   }
 
+  @Override
+  public void populateDisplayData(DisplayData.Builder builder) {
+    builder
+        .add("source", source.getClass())
+        .include(source);
+
+    if (maxNumRecords != Long.MAX_VALUE) {
+      builder.add("maxRecords", maxNumRecords);
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/BoundedReadFromUnboundedSource.java]: don't drop the `Num`?

Do you prefer English or Java names for keys. Could be "Maximum number of records returned", or "maxNumRecords".

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/CompressedSource.java [hunk 12]
```diff
@@ -319,6 +320,13 @@ public final boolean producesSortedKeys(PipelineOptions options) throws Exceptio
     return sourceDelegate.producesSortedKeys(options);
   }
 
+  @Override
+  public void populateDisplayData(DisplayData.Builder builder) {
+    builder
+        .include(sourceDelegate)
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/CompressedSource.java]: In `BoundedReadFromUnboundedSource` you included the "source: source.getClass()". Should we do so here as well?

--> We should be consistent as to which one we pick.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/CompressedSource.java [hunk 13]
```diff
@@ -319,6 +320,13 @@ public final boolean producesSortedKeys(PipelineOptions options) throws Exceptio
...
+        .add("compressionMode", channelFactory.toString());
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/CompressedSource.java]: not sure `.toString()` is right here. Will work for the ENUMs, but will not work well for `DecompressAccordingToFilename`. Maybe use something like `getClass().getSimpleName()`?

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubIO.java [hunk 14]
```diff
@@ -690,6 +691,33 @@ private Bound(String name, PubsubSubscription subscription, PubsubTopic topic,
...
+        if (timestampLabel != null) {
+          builder.add("timestampLabel", timestampLabel);
+        }
+
+        if (idLabel != null) {
+          builder.add("idLabel", idLabel);
+        }
+
+        if (maxNumRecords != 0) {
+          builder.add("maxRecords", maxNumRecords);
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubIO.java]: `Num`

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/Sink.java [hunk 15]
```diff
@@ -134,6 +136,16 @@
...
 
   /**
+   * {@inheritDoc}
+   *
+   * <p>By default, does not register any display data. Implementors may override this method
+   * to provide their own display metadata.
+   */
+  @Override
+  public void populateDisplayData(DisplayData.Builder builder) {
+  }
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/Sink.java]: merge to previous line.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/Source.java [hunk 16]
```diff
@@ -67,6 +69,16 @@
...
 
   /**
+   * {@inheritDoc}
+   *
+   * <p>By default, does not register any display data. Implementors may override this method
+   * to provide their own display metadata.
+   */
+  @Override
+  public void populateDisplayData(DisplayData.Builder builder) {
+  }
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/Source.java]: prev line

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java [hunk 17]
```diff
@@ -340,6 +341,14 @@ private Bound(String name, String filepattern, Coder<T> coder, boolean validate,
       }
 
       @Override
+      public void populateDisplayData(DisplayData.Builder builder) {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: Note to self: pick up review here.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java [hunk 18]
```diff
@@ -340,6 +341,14 @@ private Bound(String name, String filepattern, Coder<T> coder, boolean validate,
...
+        builder.add("compressionType", compressionType.toString());
+        if (filepattern != null) {
+          builder.add("filePattern", filepattern);
+        }
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: delegate inner source?

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java [hunk 19]
```diff
@@ -632,6 +641,21 @@ public PDone apply(PCollection<T> input) {
...
+        if (filenamePrefix != null) {
+          builder.add("fileNamePrefix", filenamePrefix);
+        }
+
+        if (filenameSuffix != null && !filenameSuffix.isEmpty()) {
+          builder.add("fileNameSuffix", filenameSuffix);
+        }
+
+        if (numShards != 0) {
+          builder.add("numShards", numShards);
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: delegate inner sink?

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/XmlSink.java [hunk 20]
```diff
@@ -220,6 +221,21 @@ public void validate(PipelineOptions options) {
     public XmlWriteOperation<T> createWriteOperation(PipelineOptions options) {
       return new XmlWriteOperation<>(this);
     }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      if (baseOutputFilename != null) {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/XmlSink.java]: delegate inner sink?

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/XmlSource.java [hunk 21]
```diff
@@ -218,6 +219,28 @@ public void validate() {
   }
 
   @Override
+  public void populateDisplayData(DisplayData.Builder builder) {
+    builder.add("filePattern", getFileOrPatternSpec());
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/XmlSource.java]: delegate inner source?

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/bigtable/BigtableIO.java [hunk 22]
```diff
@@ -261,6 +262,19 @@ public void validate(PBegin input) {
     }
 
     @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      builder.add("tableId", tableId);
+
+      if (options != null) {
+        builder.add("options", options.toString());
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/bigtable/BigtableIO.java]: Bigtable options

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/AvroIO.java (hunk 1 above)
Comment (swegner) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/AvroIO.java]: Class ends up being a common metadata type, so we support it first-class for display metadata. We'll use the class name (and `shortName`) for UI.

No, we aren't specifying a coder for each transform.

I had assumed the schema information is important-- is that true? If so, is this the best way to capture it?

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/AvroIO.java (hunk 2 above)
Comment (swegner) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/AvroIO.java]: We allow annotating other "components" with the `HasDisplayData` interface, and then you can include within a transform via:

    builder.include(mySource);

This is particularly useful if a transform is constructed with some sub-components which are a black box. If the transform knows the relevant metadata that the component is using, it may be easier for the transform to specify. Either implementation is fine.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java (hunk 6 above)
Comment (swegner) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java]: This was a common thread of feedback from other PR's as well-- I'll add some constructs to register values if not default.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java (hunk 8 above)
Comment (swegner) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/BigQueryIO.java]: There's been some discussion here. Prettification would definitely be useful, but it's not on our immediate plans. We'll likely look at usage to determine what "prettification" to prioritize (syntax highlighting, json formatting, line wrapping, etc.)  /cc: @bjchambers @ajamato

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/BoundedReadFromUnboundedSource.java (hunk 11 above)
Comment (swegner) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/BoundedReadFromUnboundedSource.java]: We added support for some supplying a friendly key "label":

    builder.add("maxRecords", maxNumRecords)
        .withLabel("Maximum number of records returned");

.. but so far I haven't added them for any transforms. I think it's worth doing, but I'd like to discuss with @bjchambers & @ajamato first. I think consistency is desirable-- most/all transforms should have English labels, or few/none should. 

Opened [BEAM-180](https://issues.apache.org/jira/browse/BEAM-180) to track.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/AvroIO.java (hunk 1 above)
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/AvroIO.java]: In this case, the `schema` is effectively the Coder. I would not include it unless you plan to include the coder.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/AvroIO.java (hunk 2 above)
Comment (swegner) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/AvroIO.java]: It looks like the `AvroSource` is created inside the `.apply()` method, after some validation. I'd like it to be possible to call `populateDisplayData()` in any context-- like `.toString()`.

Given that, do you think it makes sense to create an `AvroSource` inside of `populateDisplayData()` in order to let it register metadata, or let `Avro.Read.Bound` continue to register it directly?

Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/AvroIO.java]: `.apply` is where all graph construction happens and it's the location we currently make all decisions like this. I don't think it's a great idea to break that abstraction, but you're right that there's a clear conflict.

I suspect the contract you'll end up wanting is one in which the display data is not valid until after `apply` has been called.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/CompressedSource.java (hunk 13 above)
Comment (swegner) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/CompressedSource.java]: GZIP and BZIP are enum members which implement the class, so their `getClass()` returns something like `com.google.cloud.dataflow.sdk.io.CompressedSource$CompressionMode$1`

I've modified it to check whether the implementing class is an enum, and if so use it's `.name()`.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java (hunk 18 above)
Comment (swegner) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: Same comment as from Avro-- the source isn't created until within `.apply()`. I'm not sure if it's worth it (or safe to) create a temporary Source just for display data.

**Comment by swegner** [comment]
Ok, I believe I've addressed the feedback thusfar. @dhalperi, mind taking another look?

**Comment by swegner** [comment]
@dhalperi this is a large PR. Let me know if I can do anything to help review.

**Comment by dhalperi** [comment]
Probably the best way is just send "leaf transforms" that actually have an effect in the first PR, and handle composites separately.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroIO.java [hunk 23]
```diff
@@ -485,23 +489,19 @@ private Read() {}
       final boolean validate;
 
       Bound(Class<T> type) {
-        this(null, null, "", 0, ShardNameTemplate.INDEX_OF_MAX, type, null, true);
+        this(null, FileNameTemplate.DEFAULT, 0, type, null, true);
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroIO.java]: Per our discussion, I thought you were going to ditch FileNameTemplate. It only applies to FileBasedSink IIRC and it will only show up in that metadata.

(If not, it should still be in a separate PR. This PR should be "pure" and only add DisplayData.)

**Comment by swegner** [comment]
Ok, I've split this off to be just primitive IO transforms; composites will be handled in branch [displaydata-io-composites](https://github.com/swegner/incubator-beam/pull/14). 

@dhalperi, can you take another look?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroSource.java [hunk 24]
```diff
@@ -294,6 +295,13 @@ public boolean producesSortedKeys(PipelineOptions options) throws Exception {
     return coder;
   }
 
+  @Override
+  public void populateDisplayData(DisplayData.Builder builder) {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroSource.java]: Q: should you be calling `super.populateDisplayData(builder)`? If so, you probably do not need to register anything here.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/XmlSource.java [hunk 25]
```diff
@@ -219,6 +219,19 @@ public void validate() {
   }
 
   @Override
+  public void populateDisplayData(DisplayData.Builder builder) {
+
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/XmlSource.java]: `super` and then remove some of the duplicates below?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/XmlSink.java [hunk 26]
```diff
@@ -221,6 +222,14 @@ public void validate(PipelineOptions options) {
     public XmlWriteOperation<T> createWriteOperation(PipelineOptions options) {
       return new XmlWriteOperation<>(this);
     }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+      builder
+          .addIfNotNull("rootElement", rootElementName)
+          .addIfNotNull("recordClass", classToBind);
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/XmlSink.java]: looks like this did follow the `super` pattern -- no action needed.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSource.java [hunk 27]
```diff
@@ -285,6 +292,8 @@ private static long getEstimatedSizeOfFilesBySampling(
             .splitIntoBundles(desiredBundleSizeBytes, options);
       }
     });
+
+
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSource.java]: remove?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java [hunk 28]
```diff
@@ -135,6 +135,28 @@ public void validate(PipelineOptions options) {}
   @Override
   public abstract FileBasedWriteOperation<T> createWriteOperation(PipelineOptions options);
 
+  @Override
+  public void populateDisplayData(DisplayData.Builder builder) {
+    String fileNamePattern = String.format("%s%s%s",
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java]: add `super` even if empty as a style guide. If `super` data is not desired, add a comment saying why. ?

(For people who copy and paste later.)

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java [hunk 29]
```diff
@@ -320,6 +321,21 @@ public final boolean producesSortedKeys(PipelineOptions options) throws Exceptio
     return sourceDelegate.producesSortedKeys(options);
   }
 
+  @Override
+  public void populateDisplayData(DisplayData.Builder builder) {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java]:
```java
// We explicitly do not include the parent data, instead we use the delegate inner source.
```

or something like that.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/BoundedReadFromUnboundedSource.java [hunk 30]
```diff
@@ -108,6 +109,15 @@ public String getKindString() {
     return "Read(" + approximateSimpleName(source.getClass()) + ")";
   }
 
+  @Override
+  public void populateDisplayData(DisplayData.Builder builder) {
+    builder
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/BoundedReadFromUnboundedSource.java]: probably similar to `CompressedSource`.

**Comment by dhalperi** [comment]
I think you missed `CountingSource`?

**Comment by swegner** [comment]
> I think you missed CountingSource?

`CountingSource` is deprecated in favor of `CountingInput`. I've added added display data for it in the next PR for composites.

**Comment by swegner** [comment]
I've addressed all feedback so far. Please take another look.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/DatastoreIO.java [hunk 31]
```diff
@@ -19,7 +19,8 @@
 
 import static com.google.api.services.datastore.DatastoreV1.PropertyFilter.Operator.EQUAL;
 import static com.google.api.services.datastore.DatastoreV1.PropertyOrder.Direction.DESCENDING;
-import static com.google.api.services.datastore.DatastoreV1.QueryResultBatch.MoreResultsType.NOT_FINISHED;
+import static com.google.api.services.datastore.DatastoreV1.QueryResultBatch.MoreResultsType
+    .NOT_FINISHED;
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/DatastoreIO.java]: revert?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/DatastoreIO.java [hunk 32]
```diff
@@ -74,7 +76,6 @@
 import java.util.Iterator;
 import java.util.List;
 import java.util.NoSuchElementException;
-
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/DatastoreIO.java]: revert?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/DatastoreIO.java [hunk 33]
```diff
@@ -387,6 +388,19 @@ public long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
...
     @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+      builder
+          .addIfNotDefault("host", host, DEFAULT_HOST)
+          .addIfNotNull("dataset", datasetId)
+          .addIfNotNull("namespace", namespace);
+
+      if (query != null) {
+        builder.add("query", query.toString());
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/DatastoreIO.java]: is `toString()` not the default implementation for the fallback method `addIfNotNull(String, Object)`?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java [hunk 34]
```diff
@@ -55,7 +56,6 @@
 import java.util.Collections;
 import java.util.LinkedList;
 import java.util.List;
-
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java]: not sure these lines are supposed to be removed according to style guide.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/ReadTest.java [hunk 35]
```diff
@@ -17,28 +17,35 @@
...
 
 /**
  * Tests for {@link Read}.
  */
 @RunWith(JUnit4.class)
-public class ReadTest {
+public class ReadTest implements Serializable{
   @Rule
-  public ExpectedException thrown = ExpectedException.none();
+  public transient ExpectedException thrown = ExpectedException.none();
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/ReadTest.java]: ick. Probably should just go ahead and switch from anonymous inner classes to `private static` inner classes to avoid complicating future work.

Optional -- up to you.

**Comment by dhalperi** [comment]
LGTM modulo comments. When tests are green and you are happy let me know to merge.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/ReadTest.java (hunk 35 above)
Comment (swegner) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/ReadTest.java]: I'm going to leave it. Anonymous classes end up being useful in many test cases. I could move this case but it seems likely we might want to use anonymous classes again in the future.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/DatastoreIO.java (hunk 33 above)
Comment (swegner) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/DatastoreIO.java]: We handle a small set of explicit types for display data (string, integer, floating point, java class, duration, timestamp, boolean), and we have `.add` overloads for the corresponding java types.

We chose not to add a catch-all for `Object` so that specifying the type for display data is deliberate. This adds some protection against inadvertently using the wrong type (`MyCustomTimestamp`), or a type which doesn't implement toString (`MyCustomStruct`).

**Comment by swegner** [comment]
@dhalperi I think we're good-to-go. There's a single travis failure, but it looks like a flake.

**Comment by dhalperi** [comment]
Digging into the flake: Maven passes all the way successfully, but Travis still somehow decides that it failed:

```
Apr 22, 2016 11:13:09 PM org.apache.beam.sdk.io.FileBasedSource expandFilePattern
INFO: Matched 1 files for pattern /home/travis/build/apache/incubator-beam/*LIC?NSE
Apr 22, 2016 11:13:10 PM org.apache.beam.sdk.runners.DirectPipelineRunner run
INFO: Final aggregator value: s5-DroppedDueToLateness:SUM(0)
Apr 22, 2016 11:13:10 PM org.apache.beam.sdk.runners.DirectPipelineRunner run
INFO: Final aggregator value: s5-DroppedDueToClosedWindow:SUM(0)
Apr 22, 2016 11:13:10 PM org.apache.beam.sdk.runners.DirectPipelineRunner run
INFO: Final aggregator value: user-s1-emptyLines:SUM(33)
Apr 22, 2016 11:13:10 PM org.apache.beam.sdk.runners.DirectPipelineRunner run
INFO: Pipeline execution complete.
pass wordcount6b
All tests PASS

travis_time:end:04b191f4:start=1461366666796148545,finish=1461366790685229219,duration=123889080674
[0K
[32;1mThe command "travis_retry travis/test_wordcount.sh" exited with 0.[0m

Done. Your build exited with 1.
```

**Comment by swegner** [comment]
Backported via https://github.com/GoogleCloudPlatform/DataflowJavaSDK/pull/218
